### PR TITLE
ROX-26604: Add layer lineage to RHEL v2 data model

### PR DESCRIPTION
**Pull request by dcaravel**
## Description
Fixes an edge case that could cause discrepancies in scan versions, packages, etc. when images have a common layer but with different parent layers.

## Testing
Unit and e2e tests added.  Test image details at https://github.com/stackrox/stackrox/pull/13558

Also manually tested. The test images below have the same top/bottom layers but different middle layers, also they have tags that represent the version of the openjdk packages within the image for comparison.

**Before fix:**
Wrong version returned for 2nd scan:
```sh
$ rctl image scan -f --image=quay.io/dcaravel/temp:jdk-17.0.11.0.9-2.el8.x86_64 2>/dev/null | jq -r '.scan.components[] | select (.name == "java-17-openjdk-headless") | .name,.version'

java-17-openjdk-headless
1:17.0.11.0.9-2.el8.x86_64
```

```sh
$ rctl image scan -f --image=quay.io/dcaravel/temp:jdk-17.0.12.0.7-2.el8.x86_64 2>/dev/null | jq -r '.scan.components[] | select (.name == "java-17-openjdk-headless") | .name,.version'

java-17-openjdk-headless
1:17.0.11.0.9-2.el8.x86_64
```

**After fix:**
Correct versions returned for both scans
```sh
$ rctl image scan -f --image=quay.io/dcaravel/temp:jdk-17.0.11.0.9-2.el8.x86_64 2>/dev/null | jq -r '.scan.components[] | select (.name == "java-17-openjdk-headless") | .name,.version'

java-17-openjdk-headless
1:17.0.11.0.9-2.el8.x86_64
```

```sh
$ rctl image scan -f --image=quay.io/dcaravel/temp:jdk-17.0.12.0.7-2.el8.x86_64 2>/dev/null | jq -r '.scan.components[] | select (.name == "java-17-openjdk-headless") | .name,.version'

java-17-openjdk-headless
1:17.0.12.0.7-2.el8.x86_64
```

Also using a[ custom go app](https://github.com/dcaravel/compare-scans) that dumps + compares scans from two different ACS instances. I compared the scan results from a default OCP installation with one cluster that had the fix and another cluster without. Initially the testing was yielding different vulnerability detections, that seemed to change as the vuln loads progressed (as well as after they were fully loaded). 

To rule out this being related to this fix, I created a new image based off `scanner:4.6.0-rc2` but added the scanner binary built in this PR (so that the embedded vulns stayed the same):

```Dockerfile
FROM quay.io/rhacs-eng/scanner:2.35.x-22-g7ecfa77951 AS withfix

# ^^ image produced by this PR

FROM quay.io/rhacs-eng/scanner:4.6.0-rc.2

COPY --from=withfix /scanner /scanner
```
After initialization complete and latest vulns loaded, re-ran all scans in both ACS instances and the result were identical except for the specific image tested above ^^ - these differences were expected because that is what this PR is fixing.

### CODE REVIEW

**Comment by dcaravel** [comment]
Current CI failures will be fixed by https://github.com/stackrox/scanner/pull/1737 - once merged will rebase this PR.

**Comment by dcaravel** [comment]
Reviewers - please share your thoughts if we should add an env variable to switch this capability off (in case unforeseen issues)

**Comment by RTann** [comment]
> Reviewers - please share your thoughts if we should add an env variable to switch this capability off (in case unforeseen issues)

Yeah I'm cool with adding an env var

**Comment by dcaravel** [comment]
New CI failure - also saw same failure in [latest commit in master](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/branch-ci-stackrox-scanner-master-merge-e2e-tests/1869555027585011712) indicating its not related to this change.

File: database/pgsql/migrations/00021_rhel_layer_lineage.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+			// The lineage column mimics the existing `layer` table. The parent_lineage column is used
+			// instead of equivalent parent_id column from the 'layer' table to avoid an extra query
+			// on insert (which would be necessary to determine the parent id).
+			`ALTER TABLE rhelv2_layer ADD COLUMN IF NOT EXISTS lineage varchar;`,
+			`ALTER TABLE rhelv2_layer ADD COLUMN IF NOT EXISTS parent_lineage varchar`,
+
+			// Create a new unique constraint that includes lineage (and drop the old constraint)
+			`ALTER TABLE rhelv2_layer ADD CONSTRAINT rhelv2_layer_hash_lineage_key UNIQUE (hash, lineage)`,
+			`ALTER TABLE rhelv2_layer DROP CONSTRAINT IF EXISTS rhelv2_layer_hash_key`,
```
[the remark this answers is not in the document]
Comment (dcaravel) [reply database/pgsql/migrations/00021_rhel_layer_lineage.go:18]: Appears so, creating table without the constraint:

```sql
CREATE TABLE IF NOT EXISTS rhelv2_layer (
	id          BIGSERIAL PRIMARY KEY,
	hash        TEXT,
	parent_hash TEXT NOT NULL DEFAULT '',
	dist        TEXT,
	cpes        TEXT[]
);
```

```
   Column    |  Type  | Collation | Nullable |                 Default                  
-------------+--------+-----------+----------+------------------------------------------
 id          | bigint |           | not null | nextval('rhelv2_layer_id_seq'::regclass)
 hash        | text   |           |          | 
 parent_hash | text   |           | not null | ''::text
 dist        | text   |           |          | 
 cpes        | text[] |           |          | 
Indexes:
    "rhelv2_layer_pkey" PRIMARY KEY, btree (id)
```
Then adding the constraint and letting Postgres decide what to name it:
```sql
ALTER TABLE rhelv2_layer ADD UNIQUE (hash);
```

```
   Column    |  Type  | Collation | Nullable |                 Default                  
-------------+--------+-----------+----------+------------------------------------------
 id          | bigint |           | not null | nextval('rhelv2_layer_id_seq'::regclass)
 hash        | text   |           |          | 
 parent_hash | text   |           | not null | ''::text
 dist        | text   |           |          | 
 cpes        | text[] |           |          | 
Indexes:
    "rhelv2_layer_pkey" PRIMARY KEY, btree (id)
    "rhelv2_layer_hash_key" UNIQUE CONSTRAINT, btree (hash)
```

Yields: `rhelv2_layer_hash_key`

[the remark this answers is not in the document]
Comment (dcaravel) [reply database/pgsql/migrations/00021_rhel_layer_lineage.go:18]: The ordering was meant to ensure no data is inserted that violates the new constraint - since there would never be a time when no constraint exists (the queries in a migration do not appear to be executed in a DB transaction by default)

It also seemed like a reasonable assumption that the `rhelv2_layer` table would be empty when these migrations run since they are run by the first(ish) scanner that connects to the DB after an upgrade / restart which clears the data from the layer tables (and the DB's init container does not populate the layer* tables).  If true then the ordering shouldn't matter.

I'm happy to change it if desired.  Thoughts?

File: database/pgsql/queries.go [hunk 2]
```diff
@@ -264,9 +264,9 @@ const (
 	deleteStaleRHELv2Vulns = `DELETE FROM vuln_v2 WHERE name = ANY($1::text[]) and package_name = ANY($2::text[]) and cpe = ANY($3::text[]) and package_module = $4;`
 
 	insertRHELv2Layer = `
-		INSERT INTO rhelv2_layer (hash, parent_hash, dist, cpes)
-		VALUES ($1, $2, $3, $4)
-		ON CONFLICT (hash) DO NOTHING;`
+		INSERT INTO rhelv2_layer (hash, parent_hash, dist, cpes, lineage, parent_lineage)
+		VALUES ($1, $2, $3, $4, $5, $6)
+		ON CONFLICT (hash, lineage) DO NOTHING;`
```
[the remark this answers is not in the document]
Comment (dcaravel) [reply database/pgsql/queries.go:269]: It is theoretically possible on hash collision.

File: database/pgsql/rhelv2_layer_test.go [hunk 3]
```diff
@@ -176,25 +177,133 @@ func TestGetRHELv2Layers(t *testing.T) {
...
+		return
+	}
+	defer datastore.Close()
+
+	// Two 'fake' images will be created, each with 3 layers, the DB will resemble:
+	// id |      hash       |   parent_hash   |  dist  |      cpes      |  lineage  | parent_lineage
+	// ----+-----------------+-----------------+--------+----------------+-----------+----------------
+	//   1 | sha256:base     |                 | rhel:8 |                |           |
+	//   2 | sha256:layer1-a | sha256:base     | rhel:8 | {cpe-a,cpe2-a} | lineage   |
+	//   3 | sha256:layer1-b | sha256:base     | rhel:8 | {cpe-b,cpe2-b} | lineage   |
```
[the remark this answers is not in the document]
Comment (dcaravel) [reply database/pgsql/rhelv2_layer_test.go:215]: Yes, the lineage is a rolling hash of the 'parent' layer digests (does not include the current layers digest), so any images that share the same 'parents' / base images would have identical lineages.

For example, the two images below would have the same 'lineage':

```dockerfile
FROM ubi9:abc

COPY /supercoolapp
```

```dockerfile
FROM ubi9:abc

COPY /boringapp
```

File: database/pgsql/migrations/00021_rhel_layer_lineage.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (RTann) [reply database/pgsql/migrations/00021_rhel_layer_lineage.go:18]: eh since we never persist and never will persist this data, it does not matter much. :shipit:

File: database/pgsql/rhelv2_layer_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (RTann) [reply database/pgsql/rhelv2_layer_test.go:214]: > does not include the current layers digest

ah ok I didn't recall that, makes sense now

**Review by RTann — APPROVED** [review]
